src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/wrapper/PKCS11.java

```diff
@@ -1653,7 +1664,7 @@ protected void finalize() throws Throwable {
 static class SynchronizedPKCS11 extends PKCS11 {
 
     SynchronizedPKCS11(String pkcs11ModulePath, String functionListName)
-            throws IOException {
+            throws IOException, PKCS11Exception {
```

Comment:
Is this going to cause compatibility issues again? Maybe the wrapper constructor should catch PKCS11Exception and wrap it in an IOException?

Reply:
`PKCS11` & `SynchronizedPKCS11` constructors are only used from `PKCS11::getInstance()`, which [already throws `PKCS11Exception` in upstream code](https://github.com/openjdk/jdk17u/blob/jdk-17.0.3-ga/src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/wrapper/PKCS11.java#L155 "src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/wrapper/PKCS11.java:155").

Both `PKCS11::getInstance()` and the constructors are part of a non-public API, only accesible by breaking encapsulation through an _illegal reflective access_ (requiring `--add-exports` in modern Java versions), as in [RH2036462: rh1991003 patch breaks sun.security.pkcs11.wrapper.PKCS11.getInstance()](https://bugzilla.redhat.com/show_bug.cgi?id=2036462).

But unlike `PKCS11::getInstance()`, direct usage of the constructors has an additional caveat: [special care needs to be taken to avoid calling `C_Initialize` more than once per native `.so`](https://github.com/rh-openjdk/jdk/blob/1e26894969d911776f60eaa557d25a2b389060da/src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/wrapper/PKCS11.java#L175-L176 "src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/wrapper/PKCS11.java:175-176"). That makes `PKCS11::getInstance()` a much better candidate for this kind of unsupported usage.

I think it's not worth wrapping the `PKCS11Exception` in an `IOException`, but we started working in the upstream proposal of this, so we can see if the same concern is raised there.